### PR TITLE
WT-7286 Avoid bucket walking when gathering checkpoint handles

### PR DESCRIPTION
**Pull request by raviprakashgiri29**
Skip walking the session's dhandle cache and connection's hash bucket if we are the checkpoint thread (via WT_SESSION_IS_CHECKPOINT()) and the checkpoint string prefix matches WT_CHECKPOINT.
Assert that the checkpoint dhandle does not exist in connections list in diagnostic mode after acquiring the lock.

### CODE REVIEW

**Review by sulabhM — APPROVED** [review]
LGTM

File: src/session/session_dhandle.c [hunk 1]
```diff
@@ -405,12 +405,26 @@ __session_find_shared_dhandle(WT_SESSION_IMPL *session, const char *uri, const c
...
+     * it is guaranteed that the handle won't exist in the shared dhandle list. Skip searching
+     * through the list, and directly insert into it.
+     */
+    if (WT_SESSION_IS_CHECKPOINT(session) && checkpoint != NULL &&
+      WT_STRING_MATCH(checkpoint, WT_CHECKPOINT, strlen(WT_CHECKPOINT))) {
+        /*
+         * Let's check that our assumption that this handle is not already there in the shared list
+         * is actually correct.
+         */
+        WT_ASSERT(session, __wt_conn_dhandle_find(session, uri, checkpoint) != 0);
```
Comment (sueloverso) [inline src/session/session_dhandle.c:419]: This call still needs the `WT_WITH_HANDLE_LIST_READ_LOCK` or it could follow changing pointers and segfault. Also I think the assert should check explicitly for `WT_NOTFOUND`. Any other error is a real error. So it should be:
`WT_WITH...(ret = wt_conn_dhandle_find(...))`
`WT_ASSERT(session, ret != WT_NOTFOUND);`

That means it isn't skipped overall though, so I think this code is going to need to be under `HAVE_DIAGNOSTIC` instead. I think the code overall would read better if the conditional were flipped and the `if` portion is NOT a checkpoint. Then the `else` would be the diagnostic code.

File: src/session/session_dhandle.c [hunk 2]
```diff
@@ -428,11 +442,21 @@ __session_get_dhandle(WT_SESSION_IMPL *session, const char *uri, const char *che
...
-    __session_find_dhandle(session, uri, checkpoint, &dhandle_cache);
-    if (dhandle_cache != NULL) {
-        session->dhandle = dhandle_cache->dhandle;
-        return (0);
+    /*
+     * If we get called as part of the checkpoint, and the handle is the checkpoint to be created,
+     * it is guaranteed that the handle won't exist in either session or the shared dhandle list.
+     * Skip searching through the session dhandle list.
+     */
+    is_checkpoint = (WT_SESSION_IS_CHECKPOINT(session) && checkpoint != NULL &&
```
Comment (sueloverso) [inline src/session/session_dhandle.c:452]: You don't need the local variable here as this is the same conditional you used above (without a local variable). Or perhaps use a local boolean in both.

File: src/session/session_dhandle.c (hunk 1 above)
Comment (raviprakashgiri29) [reply src/session/session_dhandle.c:419]: Thanks @sueloverso for the comments. Yeah, even I found some issue with the current change, so changed the status of the PR to draft. 
I will address the review comments next Monday as I am working on a Skunkwork project this week.

**Comment by raviprakashgiri29** [comment]
@sueloverso I have updated the code to match the prefix of checkpoint string and addressed review comments. It is ready for review now.

File: src/session/session_dhandle.c [hunk 3]
```diff
@@ -404,13 +404,33 @@ static int
...
-    WT_WITH_HANDLE_LIST_READ_LOCK(session,
-      if ((ret = __wt_conn_dhandle_find(session, uri, checkpoint)) == 0)
-        WT_DHANDLE_ACQUIRE(session->dhandle));
-
-    if (ret != WT_NOTFOUND)
-        return (ret);
+    /*
+     * If we get called as part of the checkpoint, and the handle is the checkpoint to be created,
+     * it is guaranteed that the handle won't exist in the shared dhandle list. Skip searching
+     * through the list, and directly insert into it.
```
Comment (sueloverso) [inline src/session/session_dhandle.c:412]: I think the use of "handle" is confusing since there is no local variable called handle and what you're really referring to is the checkpoint local variable not even the uri. I suggest rewording the first line as:
If this session is taking a checkpoint and is looking for the system checkpoint name to be created, it is guaranteed...

File: src/session/session_dhandle.c [hunk 4]
```diff
@@ -404,13 +404,33 @@ static int
...
+    } else {
+#ifdef HAVE_DIAGNOSTIC
+        /*
+         * Let's check our assumption that this handle is not already there in the shared list is
+         * actually correct.
+         */
+        WT_WITH_HANDLE_LIST_READ_LOCK(session,
+          if ((ret = __wt_conn_dhandle_find(session, uri, checkpoint)) == 0)
+            WT_ASSERT(session, ret != WT_NOTFOUND));
+#endif // HAVE_DIAGNOSTIC
```
Comment (sueloverso) [inline src/session/session_dhandle.c:432]: Remove the `// HAVE_DIAGNOSTIC` from the line.

File: src/session/session_dhandle.c [hunk 5]
```diff
@@ -404,13 +404,33 @@ static int
...
+            return (ret);
+    } else {
+#ifdef HAVE_DIAGNOSTIC
+        /*
+         * Let's check our assumption that this handle is not already there in the shared list is
+         * actually correct.
+         */
+        WT_WITH_HANDLE_LIST_READ_LOCK(session,
+          if ((ret = __wt_conn_dhandle_find(session, uri, checkpoint)) == 0)
+            WT_ASSERT(session, ret != WT_NOTFOUND));
```
Comment (sueloverso) [inline src/session/session_dhandle.c:431]: I don't think this is right. The assertion is not meaningful because as written `ret` is guaranteed to be equal to 0 and we want to assert that `ret == WT_NOTFOUND`. I think the `if` should not be part of the statement in the `WT_WITH_HANDLE_LIST_READ_LOCK`. It should look like:
`WT_WITH_HANDLE_LIST_READ_LOCK(session, ret = __wt_conn_dhandle_find(session, uri, checkpoint)))`
`WT_ASSERT(session, ret == WT_NOTFOUND);`

File: src/session/session_dhandle.c [hunk 6]
```diff
@@ -428,11 +448,21 @@ __session_get_dhandle(WT_SESSION_IMPL *session, const char *uri, const char *che
...
     WT_DECL_RET;
+    bool is_checkpoint;
 
-    __session_find_dhandle(session, uri, checkpoint, &dhandle_cache);
-    if (dhandle_cache != NULL) {
-        session->dhandle = dhandle_cache->dhandle;
-        return (0);
+    /*
+     * If we get called as part of the checkpoint, and the handle is the checkpoint to be created,
+     * it is guaranteed that the handle won't exist in either session or the shared dhandle list.
```
Comment (sueloverso) [inline src/session/session_dhandle.c:455]: Make the same comment text changes as above.

File: src/session/session_dhandle.c [hunk 7]
```diff
@@ -428,11 +448,21 @@ __session_get_dhandle(WT_SESSION_IMPL *session, const char *uri, const char *che
...
+     * Skip searching through the session dhandle list.
+     */
+    is_checkpoint = (WT_SESSION_IS_CHECKPOINT(session) && checkpoint != NULL &&
+      WT_PREFIX_MATCH(checkpoint, WT_CHECKPOINT));
+    if (!is_checkpoint) {
+        __session_find_dhandle(session, uri, checkpoint, &dhandle_cache);
+        if (dhandle_cache != NULL) {
+            session->dhandle = dhandle_cache->dhandle;
+            return (0);
+        }
```
Comment (sueloverso) [inline src/session/session_dhandle.c:465]: You should probably add in the same style assertion in diagnostic mode.

**Comment by sulabhM** [comment]
@sueloverso , the changes we have here trigger the assert in several tests. Please see my note on WT-7286.
